x-pack/plugins/observability/server/services/slo/tasks/summary_cleanup_task.ts

```diff
@@ -0,0 +1,208 @@
+import {
+  ConcreteTaskInstance,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '@kbn/task-manager-plugin/server';
+import { SortResults } from '@elastic/elasticsearch/lib/api/types';
+import { SLO_SUMMARY_DESTINATION_INDEX_PATTERN } from '../../../assets/constants';
+import { StoredSLO } from '../../../domain/models';
+import { SO_SLO_TYPE } from '../../../saved_objects';
+
+export const TASK_TYPE = 'SLO:SUMMARIES-CLEANUP-TASK';
+
+function findMissingItems(soIds: string[], sloSummaryIds: string[]) {
+  const missingSummaryIds = sloSummaryIds.filter((item) => !soIds.includes(item));
+  const missingSOIds = soIds.filter((item) => !sloSummaryIds.includes(item));
+
+  return {
+    missingSummaryIds,
+    missingSOIds,
+  };
+}
+
+const SEPARATOR = '__V__';
+
+export const getDeleteQueryFilter = (sloSummaryIdsToDelete: string[]) => {
+  return sloSummaryIdsToDelete.map((id) => {
+    const [sloId, revision] = id.split(SEPARATOR);
+    return {
+      bool: {
+        must: [
+          {
+            term: {
+              'slo.id': sloId,
+            },
+          },
+          {
+            term: {
+              'slo.revision': Number(revision),
+            },
+          },
+        ],
+      },
+    };
+  });
+};
+
+export class SloSummaryCleanupTask {
+  private abortController = new AbortController();
+  private logger: Logger;
+  private taskManager?: TaskManagerStartContract;
+  private soClient?: SavedObjectsClientContract;
+  private esClient?: ElasticsearchClient;
+
+  constructor(taskManager: TaskManagerSetupContract, logger: Logger) {
+    this.logger = logger;
+
+    taskManager.registerTaskDefinitions({
+      [TASK_TYPE]: {
+        title: 'SLO Definitions Cleanup Task',
+        timeout: '3m',
+        maxAttempts: 1,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return this.runTask();
+            },
+
+            cancel: async () => {
+              this.abortController.abort('[SLO] Definitions clean up Task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  runTask = async () => {
+    const runAt = new Date().toISOString();
+
+    if (this.soClient && this.esClient) {
+      const finder = this.soClient.createPointInTimeFinder<Pick<StoredSLO, 'id' | 'revision'>>({
+        type: SO_SLO_TYPE,
+        perPage: 1000,
+        fields: ['id', 'revision'],
+      });
+      let searchAfterKey: SortResults | undefined;
+      let soIdsToCheck: string[] = [];
+      let sloSummaryIdsToDelete: string[] = [];
+
+      for await (const response of finder.find()) {
+        const soItems = response.saved_objects.map(
+          (so) => `${so.attributes.id}${SEPARATOR}${so.attributes.revision}`
+        );
+
+        const { sloSummaryIds, searchAfter } = await this.fetchSloSummariesIds(searchAfterKey);
```

Comment:
I've tested locally, changing the size to 1 for both queries to avoid creating thousands of SLOs for this test. The problem is that whenever we finish iterating over the saved object, we ends the main loop. Therefore most of the summary documents are never queried for.
This could happen with one SLO definition (saved object) using a partition by key that yields thousands of SLO instances (summary documents): 1 SO -> 1000s summary documents.
Using a size of 1000 and 2000 for both request is not bullet proof, it would work most of the time for small customers having a few SLO definitions & SLO instances, but not for our SREs

Reply:
if you want to test locally:
Create 3 SLOs with partition by key. Then add two summary documents: one with an existing slo id but change the revision, and one with an inexistant slo id. Change the queries size to 1 for both

Both document should be deleted when the task runs.